Clarifying SVN vs version.

matajoh opened the pull request:
This PR updates `framework_svn` and `api_svn` to be named `framework_version` and `api_version`, respectively. The original names were to represent a *semantic* version number, which was confusing in this context where SVN more commonly means *security* version number. Accordingly, the `minimum_svn` and `svn` fields for fragments will remain with the existing naming scheme but their support has been expanded to allow these strings to also be monotonically increasing integers, to be more in line with expectations around SVNs.

To provide a smooth transition, support for semantically versioned fragments and the `*_svn` naming scheme remain in the framework, but in a deprecated state. As part of this, the policy engline will provide a new error when the type of SVN is different between the SVN of the fragment and the `minimum_svn` property of a fragment policy object: `fragment svn and the specified minimum are different types`.

darracott commented on pkg/securitypolicy/framework.rego:
```diff
@@ -1778,6 +1817,17 @@ allow_unencrypted_scratch := data.policy.allow_unencrypted_scratch
+policy_api_version := data.policy.api_version
+
+# deprecated
+policy_framework_version := data.policy.framework_svn
```
`policy_framework_version` will give a rego error if both `data.policy.framework_version` and `data.policy.framework_svn` are defined right? 
I assume we have code somewhere to ensure that doesn't happen?
(Ditto for policy_api_version)

matajoh replied:
Throwing an error is the correct behaviour. If a policy contains both of these, then it is by definition invalid. In this case, this will produce the nice behaviour that the policy won't even compile.

darracott commented on pkg/securitypolicy/framework.rego:
```diff
@@ -941,10 +944,21 @@ candidate_fragments := fragments {
+    semver.compare(svn, minimum_svn) >= 0
+}
+
+svn_ok(svn, minimum_svn) {
```
Just to check I understand; here if either of the blocks defined for `svn_ok()` are true then svn_ok() is true, right?

matajoh replied:
Yes, that is correct.

darracott commented on pkg/securitypolicy/framework.rego:
```diff
@@ -1325,20 +1339,45 @@ fragment_version_is_valid {
+errors["fragment svn is below the specified minimum"] {
     input.rule == "load_fragment"
     fragment_feed_matches
+    not svn_mismatch
```
Cool! I think this all makes sense

matajoh replied:
Great.

takuro-sato reviewed (commented):
There is not so much I can check yet, but LGTM and I could run the `go run .` successfully.

I have some questions:
- In `test/cri-containerd/policy-v0.1.0.rego`, still `api_svn` and `framework_svn` are used, and I didn't find equivalent test data with `framework_version` and `api_version`. Is it fine?
- There are some massive changes on files like `.rego` files under `samples/` and  `commands.json`. Are they updated by some tool?

darracott commented on pkg/securitypolicy/framework.rego:
```diff
@@ -1650,13 +1689,13 @@ error_objects := fragments {
+check_container(raw_container, framework_version) := container {
+    semver.compare(framework_version, version) < 0
     container := {
         # Base fields
```
Why are these kv pairs only present when the framework version is smaller than the current version?

matajoh replied:
So if the policy states that it is up to date with the current framework version, then we don't need to provide backwards compatibility logic (i.e. we just use the container as-is). However, if it is an older policy, then we need to go through the different properties of containers and apply default behaviour depending on how far behind the current version the policy is.

darracott commented on pkg/securitypolicy/framework.rego:
```diff
@@ -1325,20 +1339,45 @@ fragment_version_is_valid {
+    svn_ok(data[input.namespace].svn, fragment.minimum_svn)
 }
 
 fragment_version_is_valid {
```
Does this `fragment_version_is_valid` exist just as a sanity check? E.g. would you still expect everything to work if it was deleted? I assume if there's fragment in `data.metadata.issuers[input.issuer][input.feed]` with `svn_ok(...)` evaluating to `true` then the same fragment is also in  `data.policy.fragments` and it will have the same issuer and feed and also have `svn_ok(...)` evaluating to `true`.

matajoh replied:
So apropos of nothing I believe you've found a bug in the error code! The purpose of `fragment_version_is_valid` is to provide a check so that we can write an error when the fragment version is invalid. In order to test whether a fragment is valid we need to find a matching fragment policy object which is in the policy, or which was imported in another fragment (i.e. fragments can themselves contain fragments). The second instance of `fragment_version_is_valid` is supposed to check the fragments of fragments case, but is written incorrectly.

darracott commented on internal/tools/securitypolicy/README.md:
Same question as above (Is all the new content here to do with the svn change or is there something else going on?)

darracott commented on internal/tools/policyenginesimulator/samples/simple_framework/policy.rego:
Ditto: Is all the new content here to do with the svn change or is there something else going on?

darracott commented on internal/tools/policyenginesimulator/samples/simple_framework/commands.json:
Is all the new content here to do with the svn change? It doesn't look relevant, is there something else going on?

darracott commented on pkg/securitypolicy/regopolicy_test.go:
```diff
@@ -3687,23 +3686,96 @@ func Test_Rego_LoadFragment_BadFeed(t *testing.T) {
+	}
+}
+
+func Test_Rego_LoadFragment_VersionMismatch(t *testing.T) {
```
Does this set up tests for both directions of mismatch? E.g. "fragment version as semver and min version as int" vs "fragment version as int and min version as semver". 
Asking because I saw the rego policy checked each direction separately so one direction could break while the other still works.

matajoh replied:
Yes, it tests it both ways. See line 6617

matajoh commented:
> There is not so much I can check yet, but LGTM and I could run the `go run .` successfully.
> 
> I have some questions:
> 
> * In `test/cri-containerd/policy-v0.1.0.rego`, still `api_svn` and `framework_svn` are used, and I didn't find equivalent test data with `framework_version` and `api_version`. Is it fine?

So that file is explicitly included as a old, out of date policy as a test for the backwards compatibility code, so that is by design. All the other tests automatically generate policies (see all the calls to `marshalRego`) which have the latest syntax and semantics.

> * There are some massive changes on files like `.rego` files under `samples/` and  `commands.json`. Are they updated by some tool?

Yes, those files are created by the command extractor tool, which is a tool that exists in a branch whose PR was stalled. I keep it up to date, and at some point we need to get it merged.

matajoh committed:
Clarifying SVN vs. Version.

This commit updates `framework_svn` and `api_svn` to be named `framework_version` and `api_version`,
respectively. The original names were to represent a *semantic* version number, which was confusing
in this context where SVN more commonly means *security* version number. Accordingly, the
`minimum_svn` and `svn` fields for fragments will remain with the existing naming scheme but their
support has been expanded to allow these strings to also be monotonically increasing integers, to
be more in line with expectations around SVNs.

To provide a smooth transition, support for semantically versioned fragments and the `*_svn` naming
scheme remain in the framework, but in a deprecated state. As part of this, the policy engline will
provide a new error when the type of SVN is different between the SVN of the fragment and the
`minimum_svn` property of a fragment policy object: `fragment svn and the specified minimum are different types`.

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

anmaxvl commented:
changes look good to me
